src/chat.rs

```diff
@@ -1409,95 +1426,118 @@ pub(crate) async fn update_special_chat_names(context: &Context) -> Result<()> {
+                params.set_int(Param::Selftalk, 1);
+            }
+            DC_CONTACT_ID_DEVICE => {
+                params.set_int(Param::Devicetalk, 1);
+            }
+            _ => (),
+        }
 
-            transaction.execute(
-                "INSERT INTO chats_contacts
+        context
+            .sql
+            .transaction(move |transaction| {
+                transaction.execute(
+                    "INSERT INTO chats
+                     (type, name, param, blocked, created_timestamp)
+                     VALUES(?, ?, ?, ?, ?)",
+                    params![
+                        Chattype::Single,
+                        chat_name,
+                        params.to_string(),
+                        create_blocked as u8,
+                        time(),
+                    ],
+                )?;
+
+                transaction.execute(
+                    "INSERT INTO chats_contacts
                  (chat_id, contact_id)
                  VALUES((SELECT last_insert_rowid()), ?)",
-                params![contact_id],
-            )?;
-
-            Ok(())
-        })
-        .await?;
-
-    if contact_id == DC_CONTACT_ID_SELF {
-        update_saved_messages_icon(context).await?;
-    } else if contact_id == DC_CONTACT_ID_DEVICE {
-        update_device_icon(context).await?;
-    }
+                    params![contact_id],
+                )?;
 
-    lookup_by_contact_id(context, contact_id).await
-}
-
-pub(crate) async fn lookup_by_contact_id(
-    context: &Context,
-    contact_id: u32,
-) -> Result<(ChatId, Blocked)> {
-    ensure!(context.sql.is_open().await, "Database not available");
-
-    let row = context
-        .sql
-        .query_row(
-            "SELECT c.id, c.blocked
-               FROM chats c
-              INNER JOIN chats_contacts j
-                      ON c.id=j.chat_id
-              WHERE c.type=100
-                AND c.id>9
-                AND j.contact_id=?;",
-            paramsv![contact_id as i32],
-            |row| {
-                Ok((
-                    row.get::<_, ChatId>(0)?,
-                    row.get::<_, Option<_>>(1)?.unwrap_or_default(),
-                ))
-            },
-        )
-        .await?;
-    Ok(row)
-}
+                Ok(())
+            })
+            .await?;
 
-pub async fn get_by_contact_id(context: &Context, contact_id: u32) -> Result<ChatId> {
-    let (chat_id, blocked) = lookup_by_contact_id(context, contact_id).await?;
-    ensure_eq!(blocked, Blocked::Not, "Requested contact is blocked");
+        match contact_id {
+            DC_CONTACT_ID_SELF => update_saved_messages_icon(context).await?,
+            DC_CONTACT_ID_DEVICE => update_device_icon(context).await?,
+            _ => (),
+        }
 
-    Ok(chat_id)
+        Self::lookup_by_contact(context, contact_id)
+            .await
+            .and_then(|maybe| {
+                maybe.ok_or_else(|| Error::msg("Unknown failure creating ChatId for contact"))
```

Comment:
I would rather say "failure to retrieve ChatId of chat immediately after its creation". Or, even better, within transaction call [`transaction.last_insert_rowid`](https://docs.rs/rusqlite/0.25.1/rusqlite/struct.Connection.html#method.last_insert_rowid), it never fails and is faster. `Transaction` derefs to `Connection` automatically.